File: lib/engine/margin.ts
// ============================================================================
// Aegis Vault — Margin & Liquidation System
// Pure calculation functions for margin requirements and liquidation.
// ============================================================================

import {
  type TraderAccount,
  Side,
  DEFAULT_MAINTENANCE_MARGIN_BPS,
  BPS_DENOMINATOR,
} from './types';
import { calculateUnrealizedPnl } from './pnl';

/**
 * Calculate initial margin required for a position.
 * initialMargin = notional / leverage
 */
export function requiredInitialMargin(
  notional: number,
  leverage: number
): number {
  if (leverage <= 0) throw new Error('Leverage must be positive');
  return notional / leverage;
}

/**
 * Calculate maintenance margin required.
 * maintenanceMargin = notional * maintenanceMarginBps / 10000
 * Default: 5% (500 bps)
 */
export function requiredMaintenanceMargin(
  notional: number,
  maintenanceMarginBps: number = DEFAULT_MAINTENANCE_MARGIN_BPS
): number {
  return (notional * maintenanceMarginBps) / BPS_DENOMINATOR;
}

/**
 * Calculate margin health in basis points.
 * health = (collateral + unrealizedPnl) / notional * 10000
 * Higher is healthier. Below maintenance margin bps = liquidatable.
 */
export function marginHealth(
  collateral: number,
  unrealizedPnl: number,
  notional: number
): number {
  if (notional === 0) return BPS_DENOMINATOR; // No position = max health
  return ((collateral + unrealizedPnl) / notional) * BPS_DENOMINATOR;
}

/**
 * Calculate the liquidation price for a position.
 *
 * For longs:
 *   liqPrice = entry * (1 - 1/leverage + maintenanceMarginBps/10000)
 *   Simplified: liqPrice = entry * (1 - (1/leverage - maint/10000))
 *
 * For shorts:
 *   liqPrice = entry * (1 + 1/leverage - maintenanceMarginBps/10000)
 *   Simplified: liqPrice = entry * (1 + (1/leverage - maint/10000))
 */
export function liquidationPrice(
  entryPrice: number,
  side: Side,
  maintenanceMarginBps: number = DEFAULT_MAINTENANCE_MARGIN_BPS,
  leverage: number = 1
): number {
  if (leverage <= 0) throw new Error('Leverage must be positive');

  const marginFraction = 1 / leverage;
  const maintFraction = maintenanceMarginBps / BPS_DENOMINATOR;

  if (side === Side.Long) {
    // Long liquidation: price drops below entry - (margin - maintenance)
    return entryPrice * (1 - marginFraction + maintFraction);
  } else {
    // Short liquidation: price rises above entry + (margin - maintenance)
    return entryPrice * (1 + marginFraction - maintFraction);
  }
}

/**
 * Check if a trader can withdraw a given amount without going below
 * initial margin requirements.
 */
export function canWithdraw(
  account: TraderAccount,
  amount: number,
  markPrices: Map<string, number>
): boolean {
  if (amount > account.collateralBalance) return false;

  const remainingCollateral = account.collateralBalance - amount;

  // Calculate total margin required across all positions
  let totalMarginRequired = 0;
  for (const pos of account.positions) {
    const markPrice = markPrices.get(pos.market) ?? pos.entryPrice;
    const notional = pos.size * markPrice;
    const pnl = calculateUnrealizedPnl(pos, markPrice);
    // After withdrawal, need at least maintenance margin
    const maintMargin = requiredMaintenanceMargin(notional);
    totalMarginRequired += maintMargin;

    // Check health with remaining collateral
    const health = marginHealth(remainingCollateral + pnl, 0, notional);
    if (health < DEFAULT_MAINTENANCE_MARGIN_BPS) {
      return false;
    }
  }

  return remainingCollateral >= totalMarginRequired || account.positions.length === 0;
}

/**
 * Check if a position is liquidatable (health < maintenance margin).
 */
export function isLiquidatable(
  collateral: number,
  unrealizedPnl: number,
  notional: number,
  maintenanceMarginBps: number = DEFAULT_MAINTENANCE_MARGIN_BPS
): boolean {
  const health = marginHealth(collateral, unrealizedPnl, notional);
  return health < maintenanceMarginBps;
}

File: engine/src/margin.rs
use std::collections::HashMap;
use crate::pnl::calculate_unrealized_pnl;
use crate::types::{Side, TraderAccount, BPS_DENOMINATOR, DEFAULT_MAINTENANCE_MARGIN_BPS};

pub fn required_initial_margin(notional: f64, leverage: f64) -> Result<f64, &'static str> {
    if leverage <= 0.0 {
        return Err("Leverage must be positive");
    }
    Ok(notional / leverage)
}

pub fn required_maintenance_margin(notional: f64, maintenance_margin_bps: Option<f64>) -> f64 {
    let maint = maintenance_margin_bps.unwrap_or(DEFAULT_MAINTENANCE_MARGIN_BPS);
    (notional * maint) / BPS_DENOMINATOR
}

pub fn margin_health(collateral: f64, unrealized_pnl: f64, notional: f64) -> f64 {
    if notional == 0.0 {
        return BPS_DENOMINATOR;
    }
    ((collateral + unrealized_pnl) / notional) * BPS_DENOMINATOR
}

pub fn liquidation_price(
    entry_price: f64,
    side: Side,
    maintenance_margin_bps: Option<f64>,
    leverage: Option<f64>,
) -> Result<f64, &'static str> {
    let lev = leverage.unwrap_or(1.0);
    if lev <= 0.0 {
        return Err("Leverage must be positive");
    }

    let maint = maintenance_margin_bps.unwrap_or(DEFAULT_MAINTENANCE_MARGIN_BPS);
    let margin_fraction = 1.0 / lev;
    let maint_fraction = maint / BPS_DENOMINATOR;

    match side {
        Side::Long => Ok(entry_price * (1.0 - margin_fraction + maint_fraction)),
        Side::Short => Ok(entry_price * (1.0 + margin_fraction - maint_fraction)),
    }
}

pub fn can_withdraw(account: &TraderAccount, amount: f64, mark_prices: &HashMap<String, f64>) -> bool {
    if amount > account.collateral_balance {
        return false;
    }

    let remaining_collateral = account.collateral_balance - amount;
    let mut total_margin_required = 0.0;

    for pos in &account.positions {
        let mark_price = *mark_prices.get(&pos.market).unwrap_or(&pos.entry_price);
        let notional = pos.size * mark_price;
        let pnl = calculate_unrealized_pnl(pos, mark_price);
        
        let maint_margin = required_maintenance_margin(notional, None);
        total_margin_required += maint_margin;

        let health = margin_health(remaining_collateral + pnl, 0.0, notional);
        if health < DEFAULT_MAINTENANCE_MARGIN_BPS {
            return false;
        }
    }

    remaining_collateral >= total_margin_required || account.positions.is_empty()
}

pub fn is_liquidatable(
    collateral: f64,
    unrealized_pnl: f64,
    notional: f64,
    maintenance_margin_bps: Option<f64>,
) -> bool {
    let maint = maintenance_margin_bps.unwrap_or(DEFAULT_MAINTENANCE_MARGIN_BPS);
    let health = margin_health(collateral, unrealized_pnl, notional);
    health < maint
}
